# Goal of the ETL Lesson

The main goal of this ETL pipelines lesson is to take the [World Bank Project data set](https://datacatalog.worldbank.org/dataset/world-bank-projects-operations) and merge this data with the [World Bank indicator data](https://data.worldbank.org/indicator/SP.POP.TOTL). Then you'll load the merged data into a database.

In the process, you'll need to transform these data sets in different ways. And finally, you'll code an ETL pipeline to extract, transform, and load the data all in one step.

# Extracting data from a csv file

The first step in an ETL pipeline is extraction. Data comes in all types of different formats, and you'll practice extracting data from csv files, JSON files, XML files, SQL databases, and the web.

In this first exercise, you'll practice extracting data from a CSV file and then navigating through the results. You'll see that extracting data is not always a straight-forward process.

This exercise contains a series of coding questions for you to solve. If you get stuck, there is a solution file called 1_csv_exercise_solution.ipynb. You can find this solution file by going to File->Open and then clicking on the file name.

# Part 1 projects_data.csv

You'll be using the following csv files:
* projects_data.csv
* population_data.csv

As a first step, try importing the projects data using the pandas [read_csv method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). The file path is just '../data/projects_data.csv'. You can see the file if you click on File->Open in the workspace and open the data folder.

In [1]:
# TODO: import the projects_data.csv file using the pandas library
# Store the results in the df_projects variable
import pandas as pd

In [2]:
df_projects = pd.read_csv('../data/projects_data.csv')

c:\Users\DTI Digital\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (44) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [4]:
df_projects.head(2)

id regionname                                        countryname  \
0  P162228      Other                                        World;World   
1  P163962     Africa  Democratic Republic of the Congo;Democratic Re...   

  prodline                  lendinginstr lendinginstrtype  \
0       RE  Investment Project Financing               IN   
1       PE  Investment Project Financing               IN   

  envassesmentcategorycode supplementprojectflg productlinetype  \
0                        C                    N               L   
1                        B                    N               L   

  projectstatusdisplay  ... mjtheme3name mjtheme4name mjtheme5name location  \
0               Active  ...          NaN          NaN          NaN      NaN   
1               Active  ...          NaN          NaN          NaN      NaN   

  GeoLocID GeoLocName Latitude Longitude Country Unnamed: 56  
0      NaN        NaN      NaN       NaN     NaN         NaN  
1      NaN        NaN      NaN       NaN     NaN         NaN  

[2 rows x 57 columns]

Did you get a DType warning? Read about what this warning is in the [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.errors.DtypeWarning.html).

Pandas tries to figure out programatically the data type of each column (integer, float, boolean, string). In this case, pandas could not automatically figure out the data type. That is because some columns have more than one possible data types. In other words, this data is messy.

You can use the dtype option to specify the data type of each column. Because there are so many columns in this data set, you can set all columns to be strings at least for now.

Try reading in the data set again using the read_csv() method. This time, also use the option dtype=str so that pandas treats everything like a string.

In [5]:
# TODO: Read in the projects_data.csv file using the read_csv method 
# and dtype = str option
df_projects = pd.read_csv('../data/projects_data.csv', dtype=str)

In [6]:
df_projects.head(2)

id regionname                                        countryname  \
0  P162228      Other                                        World;World   
1  P163962     Africa  Democratic Republic of the Congo;Democratic Re...   

  prodline                  lendinginstr lendinginstrtype  \
0       RE  Investment Project Financing               IN   
1       PE  Investment Project Financing               IN   

  envassesmentcategorycode supplementprojectflg productlinetype  \
0                        C                    N               L   
1                        B                    N               L   

  projectstatusdisplay  ... mjtheme3name mjtheme4name mjtheme5name location  \
0               Active  ...          NaN          NaN          NaN      NaN   
1               Active  ...          NaN          NaN          NaN      NaN   

  GeoLocID GeoLocName Latitude Longitude Country Unnamed: 56  
0      NaN        NaN      NaN       NaN     NaN         NaN  
1      NaN        NaN      NaN       NaN     NaN         NaN  

[2 rows x 57 columns]

In [8]:
# TODO: count the number of null values in the data set
# HINT: use the isnull() and sum() methods
df_projects.isna().sum()

id                              0
regionname                      0
countryname                     0
prodline                        0
lendinginstr                  246
lendinginstrtype              246
envassesmentcategorycode     5811
supplementprojectflg           53
productlinetype                 0
projectstatusdisplay            4
status                          4
project_name                    0
boardapprovaldate            1504
board_approval_month         1504
closingdate                  3349
lendprojectcost               125
ibrdcommamt                     0
idacommamt                      0
totalamt                        0
grantamt                        0
borrower                     5919
impagency                    6097
url                             0
projectdoc                  18248
majorsector_percent         18248
sector1                         0
sector2                      8721
sector3                     11761
sector4                     13872
sector5       

Notice that the number 18248 shows up multiple times. There is also a countryname column with 0 missing values and a Country column with 14045 missing values. This data set clearly has some issues that will need to be solved in the transform part of the pipeline.

Next, output the shape of the data frame.

In [10]:
# TODO: output the shape of the data frame
df_projects.shape

(18248, 57)

# Part 2 population_data.csv
Next, use the pandas read_csv method to read in the population_data.csv file. The path to this file is "../data/population_data.csv". When you try to read in this data set using pandas, you'll get an error because there is something wrong with the data.


In [11]:
# TODO: read in the population_data.csv file using the read_csv() method
# Put the results in a variable called df_population
df_population = pd.read_csv('../data/population_data.csv')

ParserError: Error tokenizing data. C error: Expected 3 fields in line 5, saw 63


In [26]:
f = open("../data/population_data.csv", "r")

In [28]:
for i in range(10):
    print(f.readline())
f.close()

"Arab World","ARB","Population, total","SP.POP.TOTL","92490932","95044497","97682294","100411076","103239902","106174988","109230593","112406932","115680165","119016542","122398374","125807419","129269375","132863416","136696761","140843298","145332378","150133054","155183724","160392488","165689490","171051950","176490084","182005827","187610756","193310301","199093767","204942549","210844771","216787402","224735446","230829868","235037179","241286091","247435930","255029671","260843462","266575075","272235146","277962869","283832016","289850357","296026575","302434519","309162029","316264728","323773264","331653797","339825483","348145094","356508908","364895878","373306993","381702086","390043028","398304960","406452690","414491886",

"United Arab Emirates","ARE","Population, total","SP.POP.TOTL","92634","101078","112472","125566","138529","150362","160481","170283","183194","203820","235499","278808","332760","397174","471364","554324","646943","748117","852262","952040","1042384",

In [35]:
df_population = pd.read_csv('../data/population_data.csv',header=2)

In [36]:
df_population.head()

Country Name Country Code     Indicator Name Indicator Code       1960  \
0        Aruba          ABW  Population, total    SP.POP.TOTL    54211.0   
1  Afghanistan          AFG  Population, total    SP.POP.TOTL  8996351.0   
2       Angola          AGO  Population, total    SP.POP.TOTL  5643182.0   
3      Albania          ALB  Population, total    SP.POP.TOTL  1608800.0   
4      Andorra          AND  Population, total    SP.POP.TOTL    13411.0   

        1961       1962       1963       1964       1965  ...        2009  \
0    55438.0    56225.0    56695.0    57032.0    57360.0  ...    101453.0   
1  9166764.0  9345868.0  9533954.0  9731361.0  9938414.0  ...  28004331.0   
2  5753024.0  5866061.0  5980417.0  6093321.0  6203299.0  ...  22549547.0   
3  1659800.0  1711319.0  1762621.0  1814135.0  1864791.0  ...   2927519.0   
4    14375.0    15370.0    16412.0    17469.0    18549.0  ...     84462.0   

         2010        2011        2012        2013        2014        2015  \
0    101669.0    102053.0    102577.0    103187.0    103795.0    104341.0   
1  28803167.0  29708599.0  30696958.0  31731688.0  32758020.0  33736494.0   
2  23369131.0  24218565.0  25096150.0  25998340.0  26920466.0  27859305.0   
3   2913021.0   2905195.0   2900401.0   2895092.0   2889104.0   2880703.0   
4     84449.0     83751.0     82431.0     80788.0     79223.0     78014.0   

         2016        2017  Unnamed: 62  
0    104822.0    105264.0          NaN  
1  34656032.0  35530081.0          NaN  
2  28813463.0  29784193.0          NaN  
3   2876101.0   2873457.0          NaN  
4     77281.0     76965.0          NaN  

[5 rows x 63 columns]